airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/StatePersistence.java

```diff
@@ -0,0 +1,323 @@
+    return ctx.select(DSL.asterisk())
+        .from(STATE)
+        .where(STATE.CONNECTION_ID.eq(connectionId))
+        .fetch(getStateRecordMapper())
+        .stream().toList();
+  }
+
+  /**
+   * Build Global state
+   *
+   * The list of records can contain one global shared state that is the state without streamName and
+   * without namespace The other records should be tronslated into AirbyteStreamState
+   */
+  private StateWrapper buildGlobalState(final List<StateRecord> records) {
+    // Split the global shared state from the other per stream records
+    final Map<Boolean, List<StateRecord>> partitions = records.stream()
+        .collect(Collectors.partitioningBy(r -> r.streamName == null && r.namespace == null));
+
+    final AirbyteGlobalState globalState = new AirbyteGlobalState()
+        .withSharedState(partitions.get(Boolean.TRUE).stream().map(r -> r.state).findFirst().orElse(null))
+        .withStreamStates(partitions.get(Boolean.FALSE).stream().map(this::buildAirbyteStreamState).toList());
+
+    final AirbyteStateMessage msg = new AirbyteStateMessage()
+        .withStateType(AirbyteStateType.GLOBAL)
+        .withGlobal(globalState);
+    return new StateWrapper().withStateType(StateType.GLOBAL).withGlobal(msg);
+  }
+
+  /**
+   * Build StateWrapper for a PerStream state
+   */
+  private StateWrapper buildStreamState(final List<StateRecord> records) {
+    List<AirbyteStateMessage> messages = records.stream().map(
+        record -> new AirbyteStateMessage()
+            .withStateType(AirbyteStateType.STREAM)
+            .withStream(buildAirbyteStreamState(record)))
+        .toList();
+    return new StateWrapper().withStateType(StateType.STREAM).withStateMessages(messages);
+  }
+
+  /**
+   * Build a StateWrapper for Legacy state
+   */
+  private StateWrapper buildLegacyState(final List<StateRecord> records) {
+    return new StateWrapper()
+        .withStateType(StateType.LEGACY)
+        .withLegacyState(records.get(0).state);
+  }
+
+  /**
+   * Convert a StateRecord to an AirbyteStreamState
+   */
+  private AirbyteStreamState buildAirbyteStreamState(final StateRecord record) {
+    return new AirbyteStreamState()
+        .withStreamDescriptor(new StreamDescriptor().withName(record.streamName).withNamespace(record.namespace))
+        .withStreamState(record.state);
+  }
+
+  private static RecordMapper<Record, StateRecord> getStateRecordMapper() {
+    return record -> new StateRecord(
+        record.get(STATE.TYPE, io.airbyte.db.instance.configs.jooq.generated.enums.StateType.class),
+        record.get(STATE.STREAM_NAME, String.class),
+        record.get(STATE.NAMESPACE, String.class),
+        Jsons.deserialize(record.get(STATE.STATE_).data()));
+  }
+
+  private static io.airbyte.db.instance.configs.jooq.generated.enums.StateType convertStateType(final StateType stateType) {
+    return switch (stateType) {
```

Comment:
we have a helper called `Enums.convertTo(...)` that should allow you to avoid needing to do this conversion.